crates/pallet-grandpa-finality-verifier/src/lib.rs

```diff
@@ -130,41 +137,36 @@ pub mod pallet {
         init_params: InitializationData,
     ) -> DispatchResult {
         let InitializationData {
-            genesis_hash,
+            oldest_parent_height,
+            oldest_parent_hash,
             best_known_finalized_header,
             set_id,
         } = init_params;
         let header_decoded = C::decode_header::<T>(best_known_finalized_header.as_slice())?;
         let change =
             find_scheduled_change(&header_decoded).ok_or(Error::<T>::UnsupportedScheduledChange)?;
-        // Set the best known finalized header
-        BestKnownFinalized::<T>::insert(chain_id, best_known_finalized_header);
+        let block_height = (*header_decoded.number()).into();
+        // Set the validation point
+        ValidationCheckPoint::<T>::insert(chain_id, (block_height, best_known_finalized_header));
         let authority_set = AuthoritySet {
             authorities: change.next_authorities,
             set_id,
         };
-        LastImportedBlock::<T>::insert(chain_id, (0, genesis_hash));
         CurrentAuthoritySet::<T>::insert(chain_id, authority_set);
+        // set the oldest known parent
+        OldestKnownParent::<T>::insert(chain_id, (oldest_parent_height, oldest_parent_hash));
         Ok(())
     }
 
     pub fn validate_finalized_block<T: Config, C: Chain>(
         chain_id: T::ChainId,
         object: &[u8],
     ) -> Result<(C::Hash, C::BlockNumber), DispatchError> {
+        // basic block validation
         let block = C::decode_block::<T>(object)?;
```

Comment:
fair!

Reply:
Just feels like a ticking bomb to me that will have hard to explain consequences once we actually hit that edge case.

Reply:
> once we actually hit that edge case
we would pretty much know in advance when a chain doesn't fit into `Chain` trait. So this should not be possible.

> not all chains are guaranteed to have u32
Yes, we assume u64 here. Have not seen any chains bigger than u64 though.

But since you feel strongly about it, we can change it use generics itself